In [1]:
#IMDB==MOVIES RATING DATA

In [2]:
import numpy as np

In [3]:
import pandas as pd

In [4]:
import matplotlib.pyplot as plt

In [5]:
import seaborn as sns

In [6]:
imdb=pd.read_csv("/content/imdb_data.csv")

In [7]:
imdb

id                          primaryTitle  \
0     tt0035423                        Kate & Leopold   
1     tt0065421                        The Aristocats   
2     tt0065938                        Kelly's Heroes   
3     tt0066026                               M*A*S*H   
4     tt0066206                                Patton   
...         ...                                   ...   
3343  tt9770150                             Nomadland   
3344  tt9777666                      The Tomorrow War   
3345  tt9784798           Judas and the Black Messiah   
3346  tt9844522  Escape Room: Tournament of Champions   
3347  tt9893250                          I Care a Lot   

                             originalTitle  isAdult  runtimeMinutes  \
0                           Kate & Leopold        0             118   
1                           The AristoCats        0              78   
2                           Kelly's Heroes        0             144   
3                                  M*A*S*H        0             116   
4                                   Patton        0             172   
...                                    ...      ...             ...   
3343                             Nomadland        0             107   
3344                      The Tomorrow War        0             138   
3345           Judas and the Black Messiah        0             126   
3346  Escape Room: Tournament of Champions        0              88   
3347                          I Care a Lot        0             118   

                          genres  averageRating  numVotes     budget  \
0         Comedy,Fantasy,Romance            6.4     87925   48000000   
1     Adventure,Animation,Comedy            7.1    111758    4000000   
2           Adventure,Comedy,War            7.6     52628    4000000   
3               Comedy,Drama,War            7.4     75784    3500000   
4            Biography,Drama,War            7.9    106476   12000000   
...                          ...            ...       ...        ...   
3343                       Drama            7.3    175977    5000000   
3344      Action,Adventure,Drama            6.6    220617  200000000   
3345     Biography,Drama,History            7.4     88568   26000000   
3346     Action,Adventure,Horror            5.7     56955   15000000   
3347          Comedy,Crime,Drama            6.4    141433   14000000   

           gross        release_date              directors  
0     76019048.0   December 11, 2001          James Mangold  
1     35459543.0   December 11, 1970    Wolfgang Reitherman  
2      5200000.0     January 1, 1970        Brian G. Hutton  
3     81600000.0    January 25, 1970          Robert Altman  
4     61749765.0    February 4, 1970  Franklin J. Schaffner  
...          ...                 ...                    ...  
3343  39458207.0  September 11, 2020             Chloé Zhao  
3344  14400000.0       June 30, 2021            Chris McKay  
3345   7478009.0    February 1, 2021             Shaka King  
3346  65774490.0        July 1, 2021           Adam Robitel  
3347   1351662.0  September 12, 2020             J Blakeson  

[3348 rows x 12 columns]

In [8]:
#lets check the data is load or not,check first 5 rows.

In [9]:
imdb.head()

id    primaryTitle   originalTitle  isAdult  runtimeMinutes  \
0  tt0035423  Kate & Leopold  Kate & Leopold        0             118   
1  tt0065421  The Aristocats  The AristoCats        0              78   
2  tt0065938  Kelly's Heroes  Kelly's Heroes        0             144   
3  tt0066026         M*A*S*H         M*A*S*H        0             116   
4  tt0066206          Patton          Patton        0             172   

                       genres  averageRating  numVotes    budget       gross  \
0      Comedy,Fantasy,Romance            6.4     87925  48000000  76019048.0   
1  Adventure,Animation,Comedy            7.1    111758   4000000  35459543.0   
2        Adventure,Comedy,War            7.6     52628   4000000   5200000.0   
3            Comedy,Drama,War            7.4     75784   3500000  81600000.0   
4         Biography,Drama,War            7.9    106476  12000000  61749765.0   

        release_date              directors  
0  December 11, 2001          James Mangold  
1  December 11, 1970    Wolfgang Reitherman  
2    January 1, 1970        Brian G. Hutton  
3   January 25, 1970          Robert Altman  
4   February 4, 1970  Franklin J. Schaffner

In [12]:
#basic information about datset

In [13]:
imdb.info()  # to check the info of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3348 entries, 0 to 3347
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   id              3348 non-null   object 
 1   primaryTitle    3348 non-null   object 
 2   originalTitle   3348 non-null   object 
 3   isAdult         3348 non-null   int64  
 4   runtimeMinutes  3348 non-null   int64  
 5   genres          3348 non-null   object 
 6   averageRating   3348 non-null   float64
 7   numVotes        3348 non-null   int64  
 8   budget          3348 non-null   int64  
 9   gross           3297 non-null   float64
 10  release_date    3343 non-null   object 
 11  directors       3348 non-null   object 
dtypes: float64(2), int64(4), object(6)
memory usage: 314.0+ KB


# INSIGHTS
-- the dataset contain 3348 rows and 12 columns

--there are 6 object data type,4 integers and 2 float columns.

-- is adult column showing integer,we need to convert this yes and no format to analyze the data properly.

-- release data columns is an object,which should be convert to date time format or we seperate the data ,year and month for better analysis.

In [14]:
#lets check for the missing values.

In [15]:
imdb.isnull().sum()

id                 0
primaryTitle       0
originalTitle      0
isAdult            0
runtimeMinutes     0
genres             0
averageRating      0
numVotes           0
budget             0
gross             51
release_date       5
directors          0
dtype: int64

INSIGHTS
-- gross(51 missing values)

-- gross represents total earning of a movies missing values may indicates that data was unavailable or the movie is underperformeed.

--possible solution : 1)missing values replace with mean or median
                      2)drop rows with missing values-51 rows will be dropped from the dataset.
                      3)if budget is availabl,then we estimate the gross using a median.

  -- missing dates--incomplete record.

  --possible solution--drop the rows.
  

In [16]:
# lets replace the missing values of gross with KNN imputation. ( knn is a part of machine learning )

THEORY :----
 KNN means taking  a nearest values

 -- if suppose we have two data so knn will take the nearest values of them.

In [18]:
# import the libraries for KNN

In [19]:
from sklearn.impute import KNNImputer

In [22]:
imputer=KNNImputer(n_neighbors=5)   # we can change the neghbour number

In [24]:
imdb[["budget","gross"]]=imputer.fit_transform(imdb[["budget","gross"]])

In [25]:
imdb.isnull().sum()

id                0
primaryTitle      0
originalTitle     0
isAdult           0
runtimeMinutes    0
genres            0
averageRating     0
numVotes          0
budget            0
gross             0
release_date      5
directors         0
dtype: int64

In [26]:
#drop the release date rows

In [30]:
imdb=imdb.dropna(subset=["release_date"]) #dropna =na is null value , subset =use for check na values and drop rows ,(not the entire column)

In [31]:
imdb.shape

(3343, 12)

# insights
-- after succesfullly removing the null values we have clean dataset.

-- total 3343 rows and 12 columns.

-- so lets start to do

***## IMDB EDA QUE :---- ***

# Basic Data Exploration

1)How many rows and columns are there in the dataset?

2)What are the column names, and what data types do they have?

3)Are there any missing values in the dataset? How would you handle them?

4)What is the summary of numerical columns (e.g., mean, median, min, max)?

5)How many unique movie titles are in the dataset?


6)How would you handle outliers in the budget and box_office columns?

7)Can you create a new column profit = box_office - budget?

8)How would you find the top 5 highest-rated movies?

9)What is the average rating of movies by genre?

10)How many movies were released each year? (Hint: Use groupby on the release_year column)

11)What is the total revenue generated by each production company?

12)Which director has made the most number of movies?

#Visualization & Insights

13)Can you create a histogram of movie ratings?

14)How would you visualize the trend of average ratings over the years?

15)Which genre has the highest average box office collection?

16)Create a bar chart to show the top 10 highest-grossing movies.

17)Can you plot a scatter plot between budget and box_office to analyze the correlation?

18)How can you create a bar chart for the top 10 highest-grossing movies?

19)How would you plot a boxplot of movie ratings grouped by genre?

#Advanced Analysis

20)How would you identify outliers in the budget column?

21)How would you recommend movies based on genre and rating?

22)Can you detect patterns in movies that won Oscars?

23)How would you identify and analyze trends in the most successful directors?
